src/renderer/canvas/label.ts

```diff
@@ -58,40 +72,160 @@ export class Label {
   }
 }
 
-export const drawLabel = (context: CanvasRenderingContext2D, label: Label) => {
+export const drawLabel = (context: CanvasRenderingContext2D, label: Label, type: 'node' | 'edge') => {
   const isDrawable = label.textLines.length > 0 && label.fontSize > 0;
   if (!isDrawable || !label.position) {
     return;
   }
 
-  drawTextBackground(context, label);
-  drawText(context, label);
+  const borderPoints: BorderPoint[] | undefined = drawTextBackgroundAndBorder(context, label, type);
+
+  if (!borderPoints) {
+    return;
+  }
+
+  drawTextBackground(context, borderPoints, label);
+  drawTextBackgroundBorder(context, borderPoints, label);
+  drawText(context, borderPoints, label, type);
 };
 
-const drawTextBackground = (context: CanvasRenderingContext2D, label: Label) => {
-  if (!label.properties.fontBackgroundColor || !label.position) {
+const drawTextBackgroundAndBorder = (
+  context: CanvasRenderingContext2D,
+  label: Label,
+  type: 'node' | 'edge',
+): BorderPoint[] | undefined => {
+  if (!label.position) {
     return;
   }
 
-  context.fillStyle = label.properties.fontBackgroundColor.toString();
   const margin = label.fontSize * FONT_BACKGROUND_MARGIN;
-  const height = label.fontSize + 2 * margin;
+
   const lineHeight = label.fontSize * FONT_LINE_SPACING;
+
   const baselineHeight = label.textBaseline === LabelTextBaseline.MIDDLE ? label.fontSize / 2 : 0;
 
+  const borderPoints: BorderPoint[] = [];
+
   for (let i = 0; i < label.textLines.length; i++) {
     const line = label.textLines[i];
-    const width = context.measureText(line).width + 2 * margin;
-    context.fillRect(
-      label.position.x - width / 2,
-      label.position.y - baselineHeight - margin + i * lineHeight,
-      width,
-      height,
-    );
+
+    let width = context.measureText(line).width + 2 * margin;
+    let x = label.position.x - width / 2;
+    let y = label.position.y - baselineHeight - margin + i * lineHeight;
+    let height = label.fontSize + 2 * margin;
+
+    if (label.properties.fontBackgroundPadding) {
+      width += label.properties.fontBackgroundPadding * 2;
+      height += label.properties.fontBackgroundPadding * 2;
+      x -= label.properties.fontBackgroundPadding;
+
+      if (type === 'node') {
+        y += label.properties.fontBackgroundPadding * 2 * i;
+      } else {
+        if (i === 0) {
+          y -= label.properties.fontBackgroundPadding;
+        } else {
+          y += label.properties.fontBackgroundPadding;
+        }
+      }
+    }
+
+    borderPoints.push({ x, y, width, height });
+  }
+
+  return borderPoints;
+};
+
+const calculateBorderRadius = (borderRadius: number, width: number, height: number) => {
+  // ensure that the radius isn't too large for x
+  if (width - 2 * borderRadius < 0) {
+    borderRadius = width / 2;
+  }
+
+  // ensure that the radius isn't too large for y
+  if (height - 2 * borderRadius < 0) {
+    borderRadius = height / 2;
+  }
+
+  return borderRadius;
+};
```

Comment:
This is a cool function that could go to `shapes.ts` utility file, like `getMaxValidBorderRadius`.